### ISSUE TITLE
[Bug] Multi-node TP rank-divergence deadlock: one rank wedges in NCCL proxy append (logits all-gather), peer idles at request broadcast — DeepSeek-V4 + DSpark on 2× DGX Spark (GB10)

### ISSUE THREAD

**Issue by MiaAI-Lab** [sgl-project/sglang#33289]
## Summary

Running DeepSeek-V4-Flash-0731 with `--speculative-algorithm DSPARK` across **2 nodes (TP=2)**, the server intermittently deadlocks minutes to hours into normal OpenAI-compatible traffic. Captured live, the two scheduler ranks sit in a permanent split:

- **Rank A (busy)**: spinning in the NCCL host enqueue path (`sched_yield` in `ncclLocalOpAppend` / `SaveProxy`) while posting an **all-gather of logits** — the proxy thread no longer consumes ops.
- **Rank B (peer)**: idle, parked at `_broadcast_reqs_across_ranks` — it has exited the batch/step loop and will never join that collective.

The unmatched collective wedges the NCCL proxy, the GPU stream never completes, and the scheduler watchdog eventually fires and kills the server. The failure shape matches the family described in #27221 (per-rank divergence in batch composition → mismatched collectives → NCCL deadlock), but with request timeouts **disabled** (both `SGLANG_REQ_WAITING_TIMEOUT` and `SGLANG_REQ_RUNNING_TIMEOUT` at default `-1`), so the divergence trigger here is a different, still-unidentified per-rank decision — all captured incidents have the DSpark verify path in the wedged rank's stack.

## Environment

- **SGLang**: v0.5.16 (`lmsysorg/sglang:latest`, pulled 2026-08-02, digest `sha256:7b6a35df…`)
- **Model**: `deepseek-ai/DeepSeek-V4-Flash-0731` (FP4 MoE + FP8 dense, bundled DSpark draft head)
- **Hardware**: 2× NVIDIA DGX Spark (GB10, SM121, aarch64, 128 GB unified memory each), TP=2 across nodes
- **Interconnect**: ConnectX-7 RoCEv2 between the nodes (12.9 GB/s measured NCCL all-reduce, zero error counters)
- **NCCL**: 2.28.9 (same in the reference vLLM stack, which is stable on identical hardware)
- **Key flags**: `--speculative-algorithm DSPARK --moe-runner-backend flashinfer_mxfp4 --kv-cache-dtype fp8_e4m3 --mem-fraction-static 0.85 --cuda-graph-backend-decode=disabled --watchdog-timeout 120`
- Local patches (bind-mounted, needed for DSpark on SM120/121): topk-192→512 index padding in `flash_mla_sm120.py` (the SM120 sparse-MLA kernel lacks a topk=192 instantiation — see flashinfer#3828) and a referenced-page bitmap for the pbs=256→64 page-split copy. Both wedges below also reproduce **without** these patches.

## Observed behavior

1. Server runs fine for minutes–hours, then a request stops making progress mid-decode.
2. No decode batches in the log; no error; one rank's scheduler pegged, the other idle.
3. Watchdog fires (`Scheduler watchdog timeout`), `kill_process_tree`, containers exit(0).
4. Repeatable across **five independent incidents**, under decode CUDA graphs both **on and off**.

### Incident 1 — worker (TP1) wedged, head (TP0) idle

TP1 py-spy (native, active):

```
sched_yield
ncclLocalOpAppend (proxy.cc:499)
SaveProxy (proxy.cc:580)
ncclProxySaveOp (proxy.cc:597)
uploadProxyOps (enqueue.cc:1293)
ncclLaunchKernelAfter_NoCuda (enqueue.cc:1670)
doLaunches / groupLaunch / ncclGroupEndInternal (group.cc)
ncclAllGather (collectives.cc:91)
c10d::ProcessGroupNCCL::_allgather_base
all_gather_into_tensor (distributed_c10d.py:4186)
tensor_model_parallel_all_gather (distributed/communication_op.py:47)
__call__ (triton_symm_mem_ag.py:515)
_get_logits (logits_processor.py:710)
forward (logits_processor.py:450)
forward (deepseek_v4.py:2530)
_execute_extend (runner/eager_runner.py:332)
...
_forward_prepared_verify (dspark_components/dspark_verify.py:271)
_forward_decode (dspark_components/dspark_worker_v2.py:590)
run_batch (scheduler.py:3351)
```

TP0 py-spy (idle):

```
broadcast (torch/distributed/distributed_c10d.py:2974)
broadcast_pyobj (utils/common.py:2175)
_broadcast_reqs_across_ranks (scheduler_components/request_receiver.py:198)
recv_requests (scheduler_components/request_receiver.py:90)
event_loop_overlap (scheduler.py:1570)
```

### Incident 2 — mirrored: head (TP0) wedged, worker (TP1) idle

Identical frames, roles swapped: TP0 active in `ncclLocalOpAppend` during `ncclAllGather` via `_get_logits` → `_execute_extend` → DSpark verify; TP1 idle at `_broadcast_reqs_across_ranks`.

Earlier incidents (before decode graphs were disabled) showed the downstream projection of the same wedge: the scheduler parked in `process_batch_result_decode → torch.cuda.synchronize` waiting for a stream that never completes.

## What we ruled out experimentally

- **NCCL fabric / config**: all-reduce microbenchmarks clean (12.9 GB/s, 14 µs at 2 KB, no stalls); zero RoCE error/retransmit counters on both nodes; per-node RoCEv2 GID auto-resolved correctly; env matched to a known-good vLLM stack on the same pair (`CROSS_NIC=1`, RoCEv2, no plugins). Same NCCL 2.28.9 in both engines. `NCCL_PROTO=Simple` and `NCCL_LAUNCH_MODE=PARALLEL` — wedge still recurs.
- **CUDA graphs**: recurs with `--cuda-graph-backend-decode=disabled` (prefill graphs are auto-disabled for this model anyway).
- **Attention backend / kernels**: recurs on `dsv4` (flashinfer SM120 sparse-MLA) and on the triton SM120 fallback.
- **DSpark ragged/confidence scheduling**: `SGLANG_RAGGED_VERIFY_MODE` at default `static` (budget planner not created); `SGLANG_SIMULATE_ACC_LEN` unset.
- **Request timeout aborts (#27221's trigger)**: both env timeouts at default `-1`.
- **Client aborts**: 35 streaming requests cancelled mid-prefill and mid-decode in a loop — no wedge. Normal completed requests also pass for hours under synthetic load.
- **Host memory**: recurs at `mem-fraction-static 0.85` with tens of GB free on both nodes.

The remaining common factor in all live captures: **multi-node TP=2 + DSpark verify path** in the wedged rank's stack, and the peer rank having already moved on to the inter-step request broadcast — i.e., a one-step divergence in batch composition between the per-rank schedulers.

## Expected behavior

Both scheduler ranks agree on batch composition every step; no unmatched collectives; no proxy wedge.

## Notes

- A reference vLLM stack (same two machines, same checkpoint, DSpark) is stable under the same client traffic for days, consistent with vLLM's driver-broadcast step table not having independent per-rank scheduler loops.
- Happy to provide full container logs (NCCL `PROXY,NET,INIT` INFO level from both nodes), the watchdog scheduler-debug dumps, and to test patches. If there is a debug flag that makes the two ranks log their batch composition per step (`cur_batch_for_debug`-style), a run with that enabled should catch the divergent step in the act.

**Comment by efschu** [comment]
The two stacks put one rank inside a collective and the peer back at
`_broadcast_reqs_across_ranks`, which places the divergence *before* the
collective rather than in it — a rank-local decision taken while the peer
decided otherwise. That family is hard to catch after the fact because
nothing records what each rank decided, only where it ended up.

One structural note on the artifacts you already have. The scheduler
watchdog's `dump_info` prints `cur_batch_for_debug` for the rank it runs on,
and its `is_active` predicate is `scheduler.is_initializing or
scheduler.cur_batch_for_debug is not None`. On the idle peer
`cur_batch_for_debug` is `None`, so that watchdog is inactive and never
dumps — the rank whose decision you actually need is the one that cannot
report. That may explain why the dumps have not localized it.

What worked for us on the same failure shape (four occurrences, always a
rank-local condition evaluated before a group collective) was a per-rank
ordered recorder rather than a snapshot: append `(op, nbytes, path)` at every
collective dispatch into a bounded in-process ring, plus one JSON line per
decision written to a per-rank file and flushed immediately. The flush per
decision is the load-bearing part — a wedged rank never gets to flush a
buffer, and the entries that matter are the last few before the wedge. A pure
offline comparator then returns the first index at which two ranks' sequences
differ, keyed on a total counter so two logs that started at different points
are not compared as if aligned. Off by default behind one env var; when off
the hot path costs a single module-global boolean test. The value is that it
names the first divergent decision instead of the terminal state.

Applied to your capture, the interesting record is not the all-gather that
hung but whatever the peer did or did not enqueue one step earlier. If a
maintainer wants it, the same recorder can be scoped to the scheduler loop
alone — `(step, forward_mode, batch_size, num_tokens)` per rank — which is
the `cur_batch_for_debug`-style flag you asked for, made ordered.

**Comment by MiaAI-Lab** [comment]
This is exactly the missing piece — thank you. The detail about the idle peer's watchdog being inactive (`cur_batch_for_debug is None` → `is_active` False) explains why every dump we captured was from the rank that already knew the answer.

We've implemented the ordered recorder you described, scoped to the scheduler loop: one JSON line per `run_batch` decision per rank — `(forward_ct, ts, forward_mode, batch_size, num_reqs, rid-suffixes)` — appended to a per-rank file, line-buffered with an explicit flush per entry, gated behind one env var (`SGLANG_RANK_RECORDER`, default off). `forward_ct` serves as the total counter for cross-rank alignment.

It's live on the 2-node setup now. The next time the wedge reproduces (it has been happening every few hours under agentic traffic), we'll have both ranks' ordered decision logs and can diff for the first divergent step. Will post the divergent record here when caught.

If useful, happy to send the ~25-line patch as a PR — when off, the hot path costs a single env-guarded boolean test.

**Comment by JackZeng0208** [comment]
Upgrade NCCL to newest version (2.30.7) and it will solve it

**Comment by roha-russ** [comment]
We independently reproduced a very similar hang on 2× GB10 (aarch64), SGLang v0.5.17, DeepSeek-V4-Flash-0731-NVFP4, TP=2, and NCCL 2.28.9.

Initially, the failure did not look obviously NCCL-specific. We reproduced it with CUDA graphs disabled, overlap scheduling disabled, and even with IB/RoCE disabled using the NCCL socket transport. The rank that wedged also changed between runs, so it did not appear to be tied to a particular node. We also tested some SGLang/DeepSeek-specific paths such as MHC post-processing without eliminating the hang.

The useful clue came from capturing both ranks live with GDB.

The wedged rank was spinning in:

`ncclLocalOpAppend -> SaveProxy(type=0 / proxyRecv)`

At that point both `proxyOps->freeOp == -1` and `pool->freeOps[0] == -1`, while the active, posted, progress-private, and producer-local queues were empty. The NCCL progress thread had nothing active to process.

On the peer rank, the proxy thread was still inside `recvProxyProgress`, waiting for network data for the same NCCL opCount. This suggested that the first rank had exhausted/lost its proxy-op descriptors before it could finish registering the corresponding send/recv work, leaving the peer waiting indefinitely.

We then found that NCCL 2.29.3 contains a very relevant ARM-specific fix in exactly this path:

https://github.com/NVIDIA/nccl/commit/25368a7f78bae866f29e46938af94fa586c84484

The release explicitly fixes weak-CAS handling in `ncclProxyGetPostedOps()` that could cause a hang on ARM. This is also the function that returns consumed proxy descriptors to `pool->freeOps[]`, so it matches the state we observed very closely.

Finally, we changed **only the runtime NCCL library** from 2.28.9 to 2.30.7. Same model, same SGLang v0.5.17, same two GB10 nodes, same TP=2 configuration and workload.

The previously failing concurrency-8 sweep completed:

* max_tokens: 1024, 2048, 4096, 8192
* 32/32 requests successful
* no hang

So in our case, upgrading NCCL appears to be the actual fix, and the per-rank divergence seen after the wedge may be a downstream consequence rather than the original trigger.

For anyone who wants to keep using the current SGLang image, this is the minimal workaround we are using:

```dockerfile
FROM lmsysorg/sglang:v0.5.17

RUN python3 -m pip install \
    --no-cache-dir \
    --force-reinstall \
    --no-deps \
    nvidia-nccl-cu13==2.30.7
```

No SGLang code changes were required.

Given that NCCL 2.29+ feature integration is still ongoing in SGLang, it may at least be useful to document or avoid NCCL 2.28.9 on aarch64/GB10, since the runtime library can be upgraded independently of those newer NCCL integrations.

**Comment by Morpheus799** [comment]
Adding an independent data point from a **different platform** that may help
separate two failure modes being discussed here.

**Setup:** single-node TP=8 on 8× H100 (x86_64), DeepSeek-V4-Flash-0731,
`--speculative-algorithm DSPARK`, HiCache + Mooncake L3 with SSD offload,
NCCL 2.28.9. Hang reproduces in ~20–60 min under high-concurrency load;
watchdog fires on `_ALLGATHER_BASE` with "wrong sizes used across ranks".

**The divergent step (captured at dispatch).** We wrapped
`torch.distributed.all_gather_into_tensor` and logged input shapes per rank.
On the wedged collective, exactly one rank fed a tensor with one fewer row:

| run | diverging rank | other ranks |
|-----|----------------|-------------|
| A   | (28, 16160)    | (29, 16160) |
| B   | (11, 16160)    | (12, 16160) |

Call site: `models/deepseek_v4_dspark.py::_apply_step_logits_sharded` →
`attn_tp_group.all_gather(step_local, dim=-1)`. That gather concatenates on the
vocab dim, so it requires an identical token dim across ranks; one rank picking
one fewer speculative token is enough to deadlock it. This matches the DSpark
per-rank sampling divergence that #33614 fixes.

**Re: NCCL 2.28.9 vs DSpark divergence.** On x86 (strong memory ordering) the
aarch64 weak-CAS proxy fix shouldn't apply, and the shapes above are measured at
the Python dispatch site — upstream of NCCL's proxy machinery — so the row-count
difference is produced by DSpark's sampling before NCCL is involved. In other
words, a genuine cross-rank size mismatch that an NCCL upgrade would not fix on
this platform. It looks like there are two distinct issues in this thread: the
aarch64 NCCL 2.28.9 proxy hang, and the platform-independent DSpark TP sampling
divergence (#33614). Disabling DSPARK removes the hang for us.

**Trigger.** It's much faster to hit under HiCache + Mooncake SSD offload:
frequent storage read failures perturb per-rank scheduler/KV timing and push the
ranks toward the sampling divergence. A pure-DRAM L3 control (no SSD offload,
same everything else) ran clean. This may be the same HiCache trigger as #32356.

Caveat: we have not isolated whether the storage *read failures specifically* or
SSD-offload timing churn more broadly is the accelerator.

**Comment by Morpheus799** [comment]
Follow-up to my data point above — I tested the NCCL-upgrade hypothesis directly on x86.

Same single-node TP=8 / 8×H100 setup, identical config, **only** the runtime NCCL
swapped 2.28.9 → **2.30.7** (`--force-reinstall nvidia-nccl-cu13==2.30.7`; confirmed
by the init banner `NCCL version 2.30.7+cuda13.3`). Nothing else changed.

**It still hangs, ~7 min into load.** Same signature:

- py-spy: all 8 ranks parked at `process_batch_result_decode → torch.unique` (the
  downstream device-sync behind the stuck forward collective).
- NCCL watchdog, `_ALLGATHER_BASE` SeqNum=65054, genuinely unequal input sizes:

  | ranks | NumelIn | = |
  |-------|---------|---|
  | rank7 | 161600  | 16160 × **10** |
  | rank0–6 | 145440 | 16160 × **9** |

  i.e. rank7 carried exactly **one extra 16160-wide vocab shard** — same ±1-unit
  divergence as my earlier runs (28 vs 29, 11 vs 12), just the odd rank on the high
  side this time. NCCL still prints "wrong sizes used across ranks".

So on x86, upgrading NCCL does **not** fix this. That's expected: a genuine cross-rank
size mismatch deadlocks on any NCCL version — NCCL is correctly refusing to gather
tensors of different shapes. The 2.29.3 aarch64 weak-CAS proxy fix can't apply here
(x86, strong memory ordering), and the divergent row count is decided by DSpark
sampling in Python, upstream of NCCL's proxy.

This reinforces that there are two separate failures in this thread: the aarch64
NCCL 2.28.9 proxy hang (fixed by the upgrade), and the platform-independent DSpark
TP sampling divergence (#33614). On x86 only #33614 / disabling DSPARK removes it —
NCCL 2.30.7 does not.

**Comment by roha-russ** [comment]
Yes, this is consistent with what we observed.

Our NCCL finding was specifically scoped to the **aarch64/GB10 proxy hang on NCCL 2.28.9**; we were not suggesting that upgrading NCCL would fix the independent DSpark cross-rank shape divergence on x86.

Thanks for taking the time to verify this distinction as well.
